File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
def print(*args, sep=" ", end="\n"): return sys.stdout.write(sep.join(map(str, args)) + end)
def input(): return sys.stdin.readline().rstrip()

def main():
    data = sys.stdin.read().splitlines()
    for s in data:
        if s == '.': break
        stack = []
        flag = 1
        for c in s:
            if c == '.': break
            if c == '(' or c == '[': stack.append(c)
            elif c == ')':
                if not stack or stack.pop() != '(': flag = 0; break
            elif c == ']':
                if not stack or stack.pop() != '[': flag = 0; break
        if flag and not stack: print('yes')
        else: print('no')
    return 0
if __name__ == '__main__':
    sys.exit(main())

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    while(true){
        string s; getline(cin, s);
        if(s == ".") break;
        stack<char> st;
        bool flag = true;
        for(char c : s){
            if(c == '('  || c == '[') st.push(c);
            else if(c == ')'){ 
                if(st.empty() || st.top() != '(') {
                    flag = false;  break;
                }
                st.pop();
            }
            else if(c == ']'){
                if(st.empty() || st.top() != '[') {
                    flag = false; break;
                }  
                st.pop(); 
            }     
        }
        if(flag && st.empty()) cout << "yes\n";
        else cout << "no\n";
    }
    return 0;
}